lib/utils/compose.js

```diff
@@ -0,0 +1,1086 @@
+		}
+		return outStream.write({
+			status,
+			progress: percentage,
+			error: errorDetail?.message ?? error,
+		});
+	};
+
+var createSpinner = function() {
+	const chars = '|/-\\';
+	let index = 0;
+	return () => chars[index++ % chars.length];
+};
+
+var runSpinner = function(tty, spinner, msg) {
+	const runloop = createRunLoop(function() {
+		tty.clearLine();
+		tty.writeLine(`${msg} ${spinner()}`);
+		return tty.cursorUp();
+	});
+	runloop.onEnd = function() {
+		tty.clearLine();
+		return tty.writeLine(msg);
+	};
+	return runloop;
+};
+
+var createRunLoop = function(tick) {
+	const timerId = setInterval(tick, 1000 / 10);
+	var runloop = {
+		onEnd() {
+			// noop
+		},
+		end() {
+			clearInterval(timerId);
+			return runloop.onEnd();
+		},
+	};
+	return runloop;
+};
+
+class BuildProgressUI {
+	constructor(tty, descriptors) {
+		this._handleEvent = this._handleEvent.bind(this);
+		this._handleInterrupt = this._handleInterrupt.bind(this);
+		this.start = this.start.bind(this);
+		this.end = this.end.bind(this);
+		this._display = this._display.bind(this);
+		const _ = require('lodash');
+		const through = require('through2');
+
+		const eventHandler = this._handleEvent;
+		const services = _.map(descriptors, 'serviceName');
+
+		const streams = _(services)
+			.map(function(service) {
+				const stream = through.obj(function(event, _enc, cb) {
+					eventHandler(service, event);
+					return cb();
+				});
+				stream.pipe(tty.stream, { end: false });
+				return [service, stream];
+			})
+			.fromPairs()
+			.value();
+
+		this._tty = tty;
+		this._serviceToDataMap = {};
+		this._services = services;
+
+		// Logger magically prefixes the log line with [Build] etc., but it doesn't
+		// work well with the spinner we're also showing. Manually build the prefix
+		// here and bypass the logger.
+		const prefix = getChalk().blue('[Build]') + '   ';
+
+		const offset = 10; // account for escape sequences inserted for colouring
+		this._prefixWidth =
+			offset + prefix.length + _.max(_.map(services, 'length'));
+		this._prefix = prefix;
+
+		// these are to handle window wrapping
+		this._maxLineWidth = null;
+		this._lineWidths = [];
+
+		this._startTime = null;
+		this._ended = false;
+		this._cancelled = false;
+		this._spinner = createSpinner();
+
+		this.streams = streams;
+	}
+
+	_handleEvent(service, event) {
+		this._serviceToDataMap[service] = event;
+	}
+
+	_handleInterrupt() {
+		this._cancelled = true;
+		this.end();
+		return process.exit(130); // 128 + SIGINT
+	}
+
+	start() {
+		process.on('SIGINT', this._handleInterrupt);
+		this._tty.hideCursor();
+		this._services.forEach(service => {
+			this.streams[service].write({ status: 'Preparing...' });
+		});
+		this._runloop = createRunLoop(this._display);
+		this._startTime = Date.now();
+	}
+
+	end(summary = null) {
+		if (this._ended) {
+			return;
+		}
+		this._ended = true;
+		process.removeListener('SIGINT', this._handleInterrupt);
+		this._runloop?.end();
```

Comment:
Added the `?` as it can technically be null although never should be in this branch